services/bots.py

```diff
@@ -30,18 +48,34 @@ def get_owner_installation_id(
 ) -> Optional[Dict]:
 
     if not ignore_installation or deprecated_using_integration:
-        default_app_installation_filter = list(
+        default_app_installation_filter: List[GithubAppInstallation] = list(
             filter(
                 lambda obj: obj.name == installation_name and obj.is_configured(),
                 owner.github_app_installations or [],
             )
         )
-        # We shuffle the results so apps are considered with close to the same probability
-        # (so that if the owner has multiple apps for the same objective their rate limit
-        # is spent equally fast)
-        shuffle(default_app_installation_filter)
+        # We assign weights to the apps based on how long ago they were updated.
+        # The idea is that there's a greater chance that a change misconfigured the app,
+        # So apps recently updated are selected less frequently than older apps
+        weights = [
+            min(MAX_GITHUB_APP_SELECTION_WEIGHT, _get_installation_weight(obj))
+            for obj in default_app_installation_filter
+        ]
+        # Random selection of size 3.
+        # If all apps have roughly the same probability of being selected, the array would have different entries.
+        # If 1 app dominates the probability of selection than it would probably be that app repeated 3 times, BUT
+        # from time to time the less frequent one would be selected.
+        apps_to_consider = (
+            random.choices(default_app_installation_filter, weights=weights, k=3)
+            if len(default_app_installation_filter) > 0
+            else []
+        )
+        already_checked = dict()
         # filter is an Iterator, so we need to scan matches
-        for app_installation in default_app_installation_filter:
+        for app_installation in apps_to_consider:
+            if already_checked.get(app_installation.installation_id):
+                continue
+            already_checked[app_installation.installation_id] = True
             if repository:
                 if app_installation.is_repo_covered_by_integration(repository):
```

Comment:
[unlikely edge case]

should the apps_to_consider also have the restriction that the repo is covered by integration, what if all the apps in the apps to consider aren't covering this repo, but there exists some app that does cover the repo in the original list of github apps?

Reply:
makes sense, it's just since the user seems to be in control of the repos that will have perms